### PR TITLE
Adding LRO implementation to msrest

### PR DESCRIPTION
**Pull request by lmazuel**

### CODE REVIEW

File: msrest/polling/poller.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+        :param int timeout: Perion of time to wait for the long running
+         operation to complete.
+        :raises CloudError: Server problem with the query.
+        """
+        if self._thread is None:
+            return
+        self._thread.join(timeout=timeout)
+        try:
+            raise self._exception
```
Comment (annatisch) [inline msrest/polling/poller.py]: This exception could be raised to the user before the thread has finished processing if the user has set a timeout - is this expected behaviour?

File: msrest/polling/poller.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        """Check status of the long running operation.
+
+        :returns: 'True' if the process has completed, else 'False'.
+        """
+        return self._thread is None or not self._thread.is_alive()
+
+    def add_done_callback(self, func):
+        """Add callback function to be run once the long running operation
+        has completed - regardless of the status of the operation.
+
```
Comment (annatisch) [inline msrest/polling/poller.py]: Maybe this comment could read: Add callback function to be run once the long running operation has completed . A callback can only be added before the operation has completed, and will be run regardless of whether the operation was successful.

File: msrest/polling/poller.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        try:
+            self._polling_method.run()
+        except Exception as err:
+            self._exception = err
+
+        finally:
+            self._done.set()
+
+        callbacks, self._callbacks = self._callbacks, []
+        while callbacks:
```
Comment (annatisch) [inline msrest/polling/poller.py]: Is this loop necessary given that we do not allow callbacks to be added or removed once self._done has been set?

File: msrest/polling/poller.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+try:
+    from urlparse import urlparse
+except ImportError:
+    from urllib.parse import urlparse
+
+from ..serialization import Model
+from ..service_client import ServiceClient
+from ..pipeline import ClientRawResponse
+
+class PollingMethod(object):
```
[the remark this answers is not in the document]
Comment (lmazuel) [reply msrest/polling/poller.py]: No, it's just convenience to give an idea of what is necessary. Duck typing :)

File: msrest/polling/poller.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    def run(self):
+        """Empty run, no polling.
+        """
+        pass
+
+    def status(self):
+        """Return the current status as a string.
+        :rtype: str
+        """
+        return "succeeded"
```
[the remark this answers is not in the document]
Comment (lmazuel) [reply msrest/polling/poller.py]: Hum, that's a good question, should I extract a status from the initial response or not. The idea was really to keep a consistent type and have the object in "result", but I guess it makes sense to make this one consistent as well. Problem is in the generic `NoPolling`, I don't know how to extract the status.
Maybe I should do a `ArmNoPolling` with status extraction (provisionningState, etc.). Thoughts?

File: msrest/polling/poller.py (hunk 2 above)
Comment (lmazuel) [reply msrest/polling/poller.py]: This should follow the asyncio behavior:
https://docs.python.org/3/library/asyncio-task.html#asyncio.Future.add_done_callback

If the future is already done when this is called, the callback is scheduled with call_soon().

File: msrest/polling/poller.py (hunk 1 above)
Comment (lmazuel) [reply msrest/polling/poller.py]: I'd yes, as soon as we have an exception let's raise it

File: msrest/polling/poller.py (hunk 2 above)
Comment (lmazuel) [reply msrest/polling/poller.py]: So I would change this code to execute the callback right now if it's done.

File: msrest/polling/poller.py (hunk 1 above)
Comment (annatisch) [reply msrest/polling/poller.py]: Yes I agree we should raise the error asap, though I wonder if the user will realize that even though an exception has been raised, self._done() may not have been set yet and any callbacks will not yet have been run. Maybe this expected/standard in which case all good - though perhaps we could add a comment to the docstring stating as such?

File: msrest/polling/poller.py (hunk 2 above)
Comment (annatisch) [reply msrest/polling/poller.py]: Just to clarify - do you mean if _done is set? Or if the thread has finished? :)

File: msrest/polling/poller.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lmazuel) [reply msrest/polling/poller.py]: So, after thinking about this, the initial call would have raised an exception is something is really wrong. Technically, the polling status of a non-polling algo is "success", so I guess you're right, we have received an initial response then the request was successful (in this context).

File: msrest/polling/poller.py (hunk 1 above)
Comment (lmazuel) [reply msrest/polling/poller.py]: Yes, let's add a comment that callbacks are executed only if the poller is done. Meaning ant timeout prevents the executions of callbacks

File: msrest/polling/poller.py (hunk 2 above)
Comment (lmazuel) [reply msrest/polling/poller.py]:
```python
        if self.done():
            func(self._polling_method)
        else:
            self._callbacks.append(func)

```

File: msrest/polling/poller.py (hunk 1 above)
Comment (annatisch) [reply msrest/polling/poller.py]: But the timeout wont prevent the execution of callbacks will it? Wont the thread continue?

File: msrest/polling/poller.py (hunk 3 above)
Comment (lmazuel) [reply msrest/polling/poller.py]: I'm not sure why this "while" was here in the first place. What kind of pattern did you implement?

Comment (lmazuel) [reply msrest/polling/poller.py]: I guess the code is trying to secure concurrent access to `self.__callbacks` and put them back after callback execution.

File: msrest/polling/poller.py (hunk 1 above)
Comment (lmazuel) [reply msrest/polling/poller.py]: Hum....... that's a good point.

Comment (lmazuel) [reply msrest/polling/poller.py]: Ok, so you confused me with your first comment :p
If `_exception` is set, this is finished anyway. So yes, it was the expected behvavior. I will revert my commit with force callbacks

**Pull request merged**